pageserver/page_api/src/client.rs

```diff
@@ -0,0 +1,234 @@
+    }
+
+    pub async fn check_rel_exists(
+        mut self,
+        req: model::CheckRelExistsRequest,
+    ) -> Result<model::CheckRelExistsResponse, tonic::Status> {
+
+        let proto_req = proto::CheckRelExistsRequest {
+            read_lsn: Some(req.read_lsn.try_into()?),
+            rel: Some(req.rel.into()),
+        };
+
+        let mut req_new = Request::new(proto_req);
+        (req_new, self) = self.insert_metadata::<proto::CheckRelExistsRequest>(req_new);
+
+        let response = self
+            .client
+            .check_rel_exists(req_new)
+            .await?;
+
+        let proto_resp = response.into_inner();
+        Ok(proto_resp.into())
+    }
+
+    pub async fn get_base_backup(
+        mut self,
+        req: &model::GetBaseBackupRequest,
+    ) -> Result<impl Stream<Item = Result<Bytes, tonic::Status>>, tonic::Status> {
+
+        let proto_req = proto::GetBaseBackupRequest {
+            read_lsn: Some(req.read_lsn.try_into()?),
+            replica: req.replica,
+        };
+
+        let mut req_new = Request::new(proto_req);
+        (req_new, self) = self.insert_metadata::<proto::GetBaseBackupRequest>(req_new);
+
+        let response_stream: Streaming<proto::GetBaseBackupResponseChunk> = self
+            .client
+            .get_base_backup(req_new)
+            .await?
+            .into_inner();
+
+        let domain_stream = response_stream.map(|chunk_res| {
+            chunk_res.and_then(|proto_chunk| {
+                let b: Bytes = proto_chunk.try_into().unwrap();
+                Ok(b)
+            })
+        });
+
+        Ok(domain_stream)
+    }
+
+    pub async fn get_db_size(
+        mut self,
+        req: model::GetDbSizeRequest,
+    ) -> Result<u64, tonic::Status> {
+
+        let proto_req = proto::GetDbSizeRequest {
+            read_lsn: Some(req.read_lsn.try_into()?),
+            db_oid: req.db_oid,
+        };
+
+        let mut req_new = Request::new(proto_req);
+        (req_new, self) = self.insert_metadata::<proto::GetDbSizeRequest>(req_new);
+
+        let response = self.client.get_db_size(req_new).await?;
+        Ok(response.into_inner().into())
+    }
+
+    pub async fn get_pages<ReqSt>(
+        mut self,
+        inbound: ReqSt,
+    ) -> Result<impl Stream<Item = Result<model::GetPageResponse, tonic::Status>> + Send + 'static, tonic::Status>
+    where
+        ReqSt: Stream<Item = model::GetPageRequest> + Send + 'static,
+    {
+        let outbound_proto: Pin<Box<dyn Stream<Item = proto::GetPageRequest> + Send>> =
+            Box::pin(
+                inbound.map(|domain_req| {
+                    domain_req.try_into().unwrap()
+                }),
+            );
+
+        let mut req_new = Request::new(outbound_proto);
+        (req_new, self) = self.insert_metadata::<Pin<Box<dyn Stream<Item = proto::GetPageRequest> + Send>>>(req_new);
+
+        let request_stream: Request<Pin<Box<dyn Stream<Item = proto::GetPageRequest> + Send>>> =
+            req_new;
```

Comment:
This is what's causing `error: implementation of `std::marker::Send` is not general enough`.

This is not clear at all, so I get why it stumped you -- async Rust has some rough edges, and the compiler errors are not as helpful as with "regular" Rust.

The problem is that the type of `inbound` gets erased when it's put inside a `Box<dyn Stream>` trait object. Because the type is erased, the compiler can't see inside it and guarantee that it's being used in a safe way inside `self.client.get_pages()`. If you don't convert it to a `Box<dyn Stream>`, but just keep the concrete types here, it'll compile.